x-pack/elastic-agent/pkg/core/server/server.go

```diff
@@ -0,0 +1,854 @@
+func New(logger *logger.Logger, listenAddr string, handler Handler) (*Server, error) {
+	ca, err := authority.NewCA()
+	if err != nil {
+		return nil, err
+	}
+	return &Server{
+		logger:                logger,
+		ca:                    ca,
+		listenAddr:            listenAddr,
+		handler:               handler,
+		apps:                  make(map[string]*ApplicationState),
+		watchdogCheckInterval: WatchdogCheckLoop,
+		checkInMinTimeout:     client.CheckinMinimumTimeout + CheckinMinimumTimeoutGracePeriod,
+	}, nil
+}
+
+// Start starts the GRPC endpoint and accepts new connections.
+func (s *Server) Start() error {
+	lis, err := net.Listen("tcp", s.listenAddr)
+	if err != nil {
+		return err
+	}
+	s.listener = lis
+	certPool := x509.NewCertPool()
+	if ok := certPool.AppendCertsFromPEM(s.ca.Crt()); !ok {
+		return errors.New("failed to append root CA", errors.TypeSecurity)
+	}
+	creds := credentials.NewTLS(&tls.Config{
+		ClientAuth:     tls.RequireAndVerifyClientCert,
+		ClientCAs:      certPool,
+		GetCertificate: s.getCertificate,
+	})
+	s.server = grpc.NewServer(grpc.Creds(creds))
+	proto.RegisterElasticAgentServer(s.server, s)
+
+	// start serving GRPC connections
+	go func() {
+		err := s.server.Serve(lis)
+		if err != nil {
+			s.logger.Errorf("error listening for GRPC: %s", err)
+		}
+	}()
+
+	// start the watchdog
+	s.watchdogDone = make(chan bool)
+	go s.watchdog()
+
+	return nil
+}
+
+// Stop stops the GRPC endpoint.
+func (s *Server) Stop() {
+	if s.server != nil {
+		close(s.watchdogDone)
+		s.server.Stop()
+		s.server = nil
+		s.listener = nil
+		s.watchdogWG.Wait()
+	}
+}
+
+// Get returns the application state from the server for the passed application.
+func (s *Server) Get(app interface{}) (*ApplicationState, bool) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	for _, appState := range s.apps {
+		if appState.app == app {
+			return appState, true
+		}
+	}
+	return nil, false
+}
+
+// Register registers a new application to connect to the server.
+func (s *Server) Register(app interface{}, config string) (*ApplicationState, error) {
+	if _, ok := s.Get(app); ok {
+		return nil, ErrApplicationAlreadyRegistered
+	}
+
+	id, err := uuid.NewV4()
+	if err != nil {
+		return nil, err
+	}
+	pair, err := s.ca.GeneratePair()
+	if err != nil {
+		return nil, err
+	}
+	appState := &ApplicationState{
+		srv:               s,
+		app:               app,
+		token:             id.String(),
+		cert:              pair,
+		pendingExpected:   make(chan *proto.StateExpected),
+		expected:          proto.StateExpected_RUNNING,
+		expectedConfigIdx: 1,
+		expectedConfig:    config,
+		checkinConn:       true,
+		status:            proto.StateObserved_STARTING,
+		statusConfigIdx:   client.InitialConfigIdx,
+		statusTime:        time.Now().UTC(),
+		pendingActions:    make(chan *pendingAction, 100),
+		sentActions:       make(map[string]*sentAction),
+		actionsConn:       true,
+	}
+	s.lock.Lock()
```

Comment:
can it be that in between get and set another set happens? in agent we do thing sync so it should be fine